test/e2e/network/dual_stack.go

```diff
@@ -0,0 +1,299 @@
+		serverDeploymentSpec := &apps.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   serverDeploymentName,
+				Labels: map[string]string{"test": "dualstack-test-server"},
+			},
+			Spec: apps.DeploymentSpec{
+				Replicas: &replicas,
+				Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"test": "dualstack-test-server"}},
+				Strategy: apps.DeploymentStrategy{
+					Type: apps.RollingUpdateDeploymentStrategyType,
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"test": "dualstack-test-server"},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "dualstack-test-server",
+								Image: imageutils.GetE2EImage(imageutils.TestWebserver),
+							},
+						},
+						Affinity: &v1.Affinity{
+							PodAntiAffinity: &v1.PodAntiAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+									{
+										LabelSelector: &metav1.LabelSelector{
+											MatchExpressions: []metav1.LabelSelectorRequirement{
+												{
+													Key:      "test",
+													Operator: metav1.LabelSelectorOpIn,
+													Values:   []string{"dualstack-test-server"},
+												},
+											},
+										},
+										TopologyKey: "kubernetes.io/hostname",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		clientDeploymentSpec := &apps.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   clientDeploymentName,
+				Labels: map[string]string{"test": "dualstack-test-client"},
+			},
+			Spec: apps.DeploymentSpec{
+				Replicas: &replicas,
+				Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"test": "dualstack-test-client"}},
+				Strategy: apps.DeploymentStrategy{
+					Type: apps.RollingUpdateDeploymentStrategyType,
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"test": "dualstack-test-client"},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:    "dualstack-test-client",
+								Image:   imageutils.GetE2EImage(imageutils.Agnhost),
+								Command: []string{"sleep", "3600"},
+							},
+						},
+						Affinity: &v1.Affinity{
+							PodAntiAffinity: &v1.PodAntiAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+									{
+										LabelSelector: &metav1.LabelSelector{
+											MatchExpressions: []metav1.LabelSelectorRequirement{
+												{
+													Key:      "test",
+													Operator: metav1.LabelSelectorOpIn,
+													Values:   []string{"dualstack-test-client"},
+												},
+											},
+										},
+										TopologyKey: "kubernetes.io/hostname",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		serverDeployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(serverDeploymentSpec)
+		framework.ExpectNoError(err)
```

Comment:
It would be preferable to add an explanation here (ref: https://github.com/kubernetes/kubernetes/issues/34059)